src/main/java/org/opensearch/security/configuration/SecurityConfigVersionHandler.java

```diff
@@ -0,0 +1,337 @@
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.security.configuration;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.action.admin.cluster.health.ClusterHealthRequest;
+import org.opensearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.support.WriteRequest.RefreshPolicy;
+import org.opensearch.cluster.health.ClusterHealthStatus;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.util.concurrent.ThreadContext;
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.security.DefaultObjectMapper;
+import org.opensearch.security.configuration.SecurityConfigVersionDocument.SecurityConfig;
+import org.opensearch.security.configuration.SecurityConfigVersionDocument.Version;
+import org.opensearch.security.securityconf.DynamicConfigFactory;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.security.securityconf.impl.SecurityDynamicConfiguration;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.user.User;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.client.Client;
+
+import org.greenrobot.eventbus.Subscribe;
+
+import static org.opensearch.security.support.ConfigConstants.SECURITY_CONFIG_VERSION_INDEX_ENABLED;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_CONFIG_VERSION_INDEX_ENABLED_DEFAULT;
+
+public class SecurityConfigVersionHandler {
+
+    private static final int MAX_VERSIONS_TO_KEEP = 10;
+
+    private static final Logger log = LogManager.getLogger(SecurityConfigVersionHandler.class);
+    private final Client client;
+    private final String SecurityConfigVersionsIndex;
+    private final SecurityConfigVersionsLoader configVersionsLoader;
+
+    private final ConfigurationRepository cr;
+    private final Settings settings;
+    private final ThreadContext threadContext;
+    private final ThreadPool threadPool;
+
+    public SecurityConfigVersionHandler(
+        ConfigurationRepository cr,
+        Settings settings,
+        ThreadContext threadContext,
+        ThreadPool threadPool,
+        Client client
+    ) {
+        this.cr = cr;
+        this.settings = settings;
+        this.threadContext = threadContext;
+        this.client = client;
+        this.SecurityConfigVersionsIndex = settings.get(
+            ConfigConstants.SECURITY_CONFIG_VERSIONS_INDEX_NAME,
```

Comment:
Do we need to add add this in OpenSearchSecurityPlugin like https://github.com/opensearch-project/security/blob/b5e7531603e281ba24f18eccd7e5ec4f6fecdc02/src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java#L2147-L2154